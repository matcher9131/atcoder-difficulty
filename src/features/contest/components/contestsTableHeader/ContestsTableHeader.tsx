import type { ReactNode } from "react";
import { useTranslation } from "react-i18next";
import { headerCellClassNames } from "../../../../common/headerCellClassNames";

export type ContestsTableHeaderProps = {
    readonly numProblems: number;
};

export const ContestsTableHeader = ({ numProblems }: ContestsTableHeaderProps): ReactNode => {
    const { t } = useTranslation();
    return (
        <thead className="contents">
            <tr className="contents">
                <th className={headerCellClassNames}>{t("contestsTableHeader.contestLabel")}</th>
                {new Array(numProblems).fill(0).map((_, i) => (
                    <th
                        key={i}
                        className={headerCellClassNames}
                    >{`${t("contestsTableHeader.problemLabel")}${i + 1}`}</th>
                ))}
            </tr>
        </thead>
    );
};
