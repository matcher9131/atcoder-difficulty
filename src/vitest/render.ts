import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
