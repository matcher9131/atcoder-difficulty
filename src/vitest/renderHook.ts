import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
