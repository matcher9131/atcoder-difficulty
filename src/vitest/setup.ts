import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
