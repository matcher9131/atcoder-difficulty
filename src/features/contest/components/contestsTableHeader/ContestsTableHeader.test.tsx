import { describe, it, expect, vi, beforeEach } from "vitest";
import { screen } from "@testing-library/react";
import { render } from "../../../../vitest/render";
import { ContestsTableHeader } from "./ContestsTableHeader";

// src/features/contest/components/contestsTableHeader/ContestsTableHeader.test.tsx

// Mock translation hook
vi.mock("react-i18next", () => ({
    useTranslation: () => ({
        t: (key: string) => {
            if (key === "contestsTableHeader.contestLabel") return "Contest";
            if (key === "contestsTableHeader.problemLabel") return "Problem";
            return key;
        },
    }),
}));

// Mock headerCellClassNames
vi.mock("../../../../common/headerCellClassNames", () => ({
    headerCellClassNames: "mock-header-cell-class",
}));

describe("ContestsTableHeader", () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("renders thead with correct class", () => {
        render(<ContestsTableHeader numProblems={3} />);
        const thead = screen.getByRole("rowgroup");
        expect(thead.tagName).toBe("THEAD");
        expect(thead.className).toContain("contents");
    });

    it("renders contest label in first header cell", () => {
        render(<ContestsTableHeader numProblems={2} />);
        const headerCells = screen.getAllByRole("columnheader");
        expect(headerCells[0]).toHaveTextContent("Contest");
    });

    it("renders correct number of problem header cells", () => {
        render(<ContestsTableHeader numProblems={4} />);
        const headerCells = screen.getAllByRole("columnheader");
        // 1 contest label + numProblems
        expect(headerCells).toHaveLength(5);
        for (let i = 1; i <= 4; i++) {
            expect(headerCells[i]).toHaveTextContent(`Problem${i.toString()}`);
        }
    });

    it("applies headerCellClassNames to all header cells", () => {
        render(<ContestsTableHeader numProblems={2} />);
        const headerCells = screen.getAllByRole("columnheader");
        headerCells.forEach((cell) => {
            expect(cell.className).toContain("mock-header-cell-class");
        });
    });

    it("matches snapshot", () => {
        const { container } = render(<ContestsTableHeader numProblems={2} />);
        expect(container).toMatchSnapshot();
    });
});
