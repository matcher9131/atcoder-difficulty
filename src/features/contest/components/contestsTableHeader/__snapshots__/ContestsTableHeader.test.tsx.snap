// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestsTableHeader > matches snapshot 1`] = `
<div>
  <thead
    class="contents"
  >
    <tr
      class="contents"
    >
      <th
        class="mock-header-cell-class"
      >
        Contest
      </th>
      <th
        class="mock-header-cell-class"
      >
        Problem1
      </th>
      <th
        class="mock-header-cell-class"
      >
        Problem2
      </th>
    </tr>
  </thead>
</div>
`;
